### ISSUE TITLE
[Enhancement] Support for scheduling using cron

### ISSUE THREAD

**Issue by BradyRosino** [AppDaemon/appdaemon#1229]
Unfortunately I don't have the skillset to contribute, but one area where I would love to see some minor improvement in AppDaemon is scheduling callbacks. I think it'd be great to have a way to schedule callbacks based on Unix cron expressions like this...

```
run_cron(self, callback, cron, **kwargs)
```

Below is an example of how I'd expect it to work where `self.some_callback` would be executed at 6:30 am Monday - Friday.

```
self.run_cron(self.some_callback,"30 6 * * 1-5")
```

Ultimately I always forget how to constrain to certain days of the week using `run_daily`. Cron is easy to hit every minute, every hour, some calendar date, only on certain days of the week, etc.

Food for thought 😀

**Comment by Odianosen25** [comment]
@BradyRosino,

Not sure its understood what you mean, at least for myself. You want a Cron based scheduler? And why will that be?

What will it give you that the present scheduler will not?

Regards

**Comment by ReneTode** [comment]
@BradyRosino 
the options are already available, but as you say you forget how to use them.
so make a bookmark from this doc section:
https://appdaemon.readthedocs.io/en/latest/APPGUIDE.html#callback-constraints

and instead of run_daily etc. you can use run_every for everything.

```
self.run_every(self.callback, "now", 60) # minutely
self.run_every(self.callback, "now", 60*60) # hourly
self.run_every(self.callback, "now", 24*60*60) # daily
self.run_every(self.callback, "now", 7*24*60*60) # weekly
self.run_every(self.callback, "now",60*60, constrain_days = "mon, tue") # hourly on some days
```

so as you see its all possible, you just want it to be written different.

**Comment by acockburn** [comment]
Rene is correct as in there would be no new functionality here, but I like the idea, and I don't think it would be too hard to implement - maybe we'll add it in the future.

**Comment by BradyRosino** [comment]
Agreed with all that there is no real new functionality being proposed. Primary benefit is that it is easier to hit some more complex time triggers using a more familiar format. I think time can be challenging for newer folks to programming and this would allow them to hit a lot of advanced datetime based triggers using one of the many available cron calculators online. AppDaemon is one of those projects that attracts many python curious people from Home Assistant. 😀

That said, @ReneTode , thank you for the link. My understanding of constrain_* was kind of warped based on samples found in Home Assistant forums. I now understand that these are constraints for _any_ callbacks and not just time based callbacks, which is certainly extremely useful.

The reason why I found this challenging is because I'd find myself at looking run_daily in the docs and there is no mention of any of the constraints available. Everything I need is available in the docs, they just (to me) seemed completely unrelated at the time I reviewed them. Now I understand. Might be worth adding a note about being able to constrain any of the run_* calls and provide an example there. With that said, I still think cron based scheduling could be useful for the reasons mentioned in my first paragraph.

Thank you all for your work on AppDaemon! 😀

**Comment by ReneTode** [comment]
> Agreed with all that there is no real new functionality being proposed. Primary benefit is that it is easier to hit some more complex time triggers using a more familiar format.

only for those who are familiar with cron. for me it wouldnt add anything because i never used cron. (and i suspect that there a a lot of HA/AD users that also never used cron.)

> I think time can be challenging for newer folks to programming and this would allow them to hit a lot of advanced datetime based triggers using one of the many available cron calculators online. AppDaemon is one of those projects that attracts many python curious people from Home Assistant.

personally i think that the examples i gave are easier to understand for everyone.
we got functions for running daily, hourly and minutely, and with run_every you can use every other amount.
i support AD for over 5 years, and i never talked to anyone who found cron easier, or mentioned that they would like it in that way.

> The reason why I found this challenging is because I'd find myself at looking run_daily in the docs and there is no mention of any of the constraints available. Everything I need is available in the docs, they just (to me) seemed completely unrelated at the time I reviewed them. Now I understand. Might be worth adding a note about being able to constrain any of the run_* calls and provide an example there.

constraints are available for all the run_... but also for listen_state and listen_event, etc.
and they are available as applevel constraints (where you place it in the yaml) or callback level constraints.
if it should be mentioned for all the possible options, with examples, maintaining would become very difficult, because it would be mentioned at a dozen places.

if you find a good way to do that, then please tell the devs, so they can make the docs better.

so to conclude: i dont think its usefull to add functions that are already there, just because that way is used in another program. if AD would start that way then there are thousands of other programs that also provide simular functions that can be copied, but AD would only become way to complex for new users.

ill leave it to Andrew and Odianosen to decide, because they are the devs (im not anymore) but they know my opinion ;)

**Comment by nsandstrom** [comment]
There is one large difference (unless I miss something), in that the execution time of run_every will depend on when the app was started.

Lets say 
`run_every(self.callback_timer, 'now', 60*15) #quarterly`

If I start this application at 14:05, it will run at 05, 20, 35, 50
If I start at 14:12, it will run at 12, 27, 42, 57

Compare that to the cron-expression `5/15 * * * *`, that will always run at 05, 20, 35, 50

**Comment by ReneTode** [comment]
then change "now" to the time you want it to start and it will run on 05,20,35, 50

the difference is that with cron you need to change the 5 in 5/15 and in run_every you need to change the starttime.

**Comment by nsandstrom** [comment]
It is of course solvable. It just takes some extra 10 lines of python to calculate a "now" value of a desired time near in the future.

I just wanted to highlight that there is an actually difference between how "now" and cron behaves.

**Comment by ReneTode** [comment]
10 lines?
```
startminute = 5
self.run_every(self.callback, datetime.datetime(dt.year, dt.month, dt.day, dt.hour,(15*(dt.minute // 15)) + startminute, 60*15)
```

**Comment by nsandstrom** [comment]
The above line will result in the previous execution (at .29 result is 15), which, being a timestamp in the past works the same way as "now".
An additional 15 min needs to be added.

For anyone interested, this code seems to be working pretty well.

```
from datetime import datetime, timedelta

startminute = 5
now = datetime.now()
start_time = datetime(
    now.year, now.month, now.day, now.hour,
    (15 * (now.minute // 15))) + timedelta(minutes=15 + startminute)
self.run_every(self.callback_timer, start_time, 15*60)
```

**Comment by ReneTode** [comment]
yeah i copy pasted something to fast.
but there are a lot of ways to do it, for example:
```
        time = self.datetime()
        startminute = 5
        starttime = self.parse_datetime("{}-{}-{} {}:{}:00".format(time.year, time.month, time.day, time.hour, (15*(time.minute // 15)))) + datetime.timedelta(minutes = 15 + startminute)
```

**Comment by nsandstrom** [comment]
I also realize that the above is not perfect.
If I start the application at say 9.02, it will execute at 9.20 and miss the 9.05 cycle.
So some conditionals will have to be added, and of done in a readable way, will probably reach those 10 lines 😉

So again, not a feature that would be absolutely necessary, but it does have it's unique use cases.

**Comment by ReneTode** [comment]
i just remembered that i already had it running for (6 or 7) years like this:
```
    nowtime = datetime.datetime.now()
    repeatminutes = self.args["minutes_between"] 
    startminute = self.args["start_minute"]
    minute = (int(nowtime.minute/repeatminutes)+1)*repeatminutes
    time = nowtime.replace(minute=startminute, second=0)+datetime.timedelta(minutes=minute)
    repeat = repeatminutes*60
    self.run_every(self.cron_job,time,repeat)
```
i use that for backups.
its possible its not perfect, but it works errorfree for me.

**Comment by nsandstrom** [comment]
It works exactly as my flawed example above.

If you run repeatminutes=15, startminute=5, and starts at 09.01, it will skip 09.05 and run at 09.20

So if your logic needs exactly 4 runs / hour, a badly timed update can cause issues.

**Comment by ReneTode** [comment]
not for me, because i start the callback immediatly as well when the app is started.
so mine would then run at 09:01 and 09:20.
also that app is only restarted about once every 6 months. and i probably will move it to an AD instance that now runs for 2 years.

but you could add a line that checks if its minute 1 and then take 15 mins off.

**Comment by nsandstrom** [comment]
And that would instead give you 5 executions the same hour.

Let's juts agree that cron and run_every behaves quite differently, run_every can be used "cron-like" with some code, but to get exactly the same behavior is a lot more complex.

**Comment by sandervandegeijn** [comment]
Would like the cron syntax as well, also running into the limitation that run_every depends on the start time of the app.
